### PR TITLE
Conditionally enable/disable section creation in `CreateSelectSectionWidget` (#284, #299)

### LINKED ISSUE #284: Create section Widget should not be shown for TA and check for Designer and Assistant
Current Non-UMICH user workflow allows create section, but TA are not allowed to create sections in Canvas so API response is 401. So we should not show Create section widget for roles that canvas don't allow

As per discussed in CCM meeting, limit the section widget access to Admins and Teachers.

### LINKED ISSUE #299: Support Consultant role is not able to unmerge sections nor can search by course name/Uniqname
This is broken at some point later when this #16 got merged

### PR DESCRIPTION
**Pull request by ssciolla**
This PR changes the `CreateSelectSectionWidget` component so that only admins (or equivalent) or teachers can see and use the section creation portion of the component.  The PR aims to resolve #284. @pushyamig also added changes to resolve #299.

### CODE REVIEW

File: ccm_web/client/src/components/CreateSelectSectionWidget.tsx [hunk 1]
```diff
@@ -28,43 +28,59 @@ const useStyles = makeStyles((theme) => ({
...
-interface CreateSelectSectionWidgetProps {
+interface WidgetWithoutCreateProps {
+  readonly canCreate: false
+}
+
+interface WidgetWithCreateProps {
+  readonly canCreate: true
   course: CanvasCourseBase
+  onSectionCreated: (section: CanvasCourseSection) => void
+}
```
Comment (pushyamig) [inline ccm_web/client/src/components/CreateSelectSectionWidget.tsx:39]: Why do you need to 2 interfaces `canCreate: true/false `? Can this be something like below? I see that you are assigning `canCreate`  true or false [here](https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/298/files#diff-c30ee72622169f2c71d21d46e688fde3c6758cd5964030732fdeda72bee0aa3fR130) when calling the Create section wizard.
 I sure there must be good reason why you are doing this. But I might not catching it. 
```
interface WidgetWithoutCreateProps {
  readonly canCreate: boolean
}
```

File: ccm_web/client/src/pages/MergeSections.tsx [hunk 2]
```diff
@@ -127,7 +127,7 @@ function MergeSections (props: CCMComponentProps): JSX.Element {
   }
 
   const isSubAccountAdminOrAccountAdmin = (): boolean => {
-    return isAuthorizedForRoles([RoleEnum['Subaccount admin'], RoleEnum['Account Admin']], props.globals.course.roles, 'MergeSections')
+    return isAuthorizedForRoles([RoleEnum['Subaccount admin'], RoleEnum['Account Admin']], props.globals.course.roles)
```
Comment (pushyamig) [inline ccm_web/client/src/pages/MergeSections.tsx:130]: Why was these removed? that seems to be right thing to do. But what did we miss?

File: ccm_web/client/src/models/feature.ts [hunk 3]
```diff
@@ -9,12 +9,14 @@ interface FeatureDataProps {
   helpURLEnding: string
 }
 
+const adminRoles = [RoleEnum['Subaccount admin'], RoleEnum['Account Admin'], RoleEnum['Support Consultant']]
+
```
Comment (pushyamig) [inline ccm_web/client/src/models/feature.ts:13]: add type `const adminRole: : RoleEnum[]`?

File: ccm_web/client/src/models/feature.ts [hunk 4]
```diff
@@ -63,10 +65,11 @@ const addNonUMUsersProps: FeatureDataProps = {
   helpURLEnding: '/add-non-um-users.html'
 }
 
-const courseRenameRoles: RoleEnum[] = [RoleEnum['Account Admin'], RoleEnum['Subaccount admin'], RoleEnum['Support Consultant']]
+const courseRenameRoles: RoleEnum[] = adminRoles
+const createSectionRoles: RoleEnum[] = [RoleEnum.Teacher, ...adminRoles]
```
Comment (pushyamig) [inline ccm_web/client/src/models/feature.ts:69]: This can going very top next to `adminRoles` constant

File: ccm_web/client/src/pages/MergeSections.tsx (hunk 2 above)
Comment (ssciolla) [reply ccm_web/client/src/pages/MergeSections.tsx:130]: The string `'MergeSections'` is not used at all, nor were the other stings. From my perspective, it's not necessary, we could always just add a log message for a particular use case if we needed it while we're developing.

File: ccm_web/client/src/components/CreateSelectSectionWidget.tsx (hunk 1 above)
Comment (ssciolla) [reply ccm_web/client/src/components/CreateSelectSectionWidget.tsx:39]: I can play around this some more, but this is the main way I know how to have the extra props `course` and `onSectionCreated` not be necessary when `canCreate === false`. We could just pass the props and not use them, but I leaned towards the component only enforcing their presence if they're needed.

File: ccm_web/client/src/models/feature.ts (hunk 3 above)
Comment (ssciolla) [reply ccm_web/client/src/models/feature.ts:13]: TypeScript already knows that is the variable's type, so I guess you are just saying to guard against future changes by adding the type?

File: ccm_web/client/src/models/feature.ts [hunk 5]
```diff
@@ -63,10 +65,11 @@ const addNonUMUsersProps: FeatureDataProps = {
...
 
 export type { FeatureDataProps }
 export {
   mergeSectionProps, formatCanvasGradebookProps, formatThirdPartyGradebookProps,
-  createSectionsProps, addUMUsersProps, addNonUMUsersProps, courseRenameRoles
+  createSectionsProps, addUMUsersProps, addNonUMUsersProps, courseRenameRoles, createSectionRoles
```
[the remark this answers is not in the document]
Comment (ssciolla) [reply ccm_web/client/src/models/feature.ts:74]: If it's not being used anywhere yet, I would lean towards not exporting it yet.

File: ccm_web/client/src/models/feature.ts (hunk 3 above)
Comment (pushyamig) [reply ccm_web/client/src/models/feature.ts:13]: Yes.

File: ccm_web/client/src/models/feature.ts (hunk 4 above)
Comment (ssciolla) [reply ccm_web/client/src/models/feature.ts]: You took care of this, resolved.

File: ccm_web/client/src/models/feature.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ssciolla) [reply ccm_web/client/src/models/feature.ts]: You're using it now, so that's okay.

File: ccm_web/client/src/models/feature.ts (hunk 3 above)
Comment (ssciolla) [reply ccm_web/client/src/models/feature.ts:15]: You added these.

File: ccm_web/client/src/components/CreateSelectSectionWidget.tsx (hunk 1 above)
Comment (ssciolla) [reply ccm_web/client/src/components/CreateSelectSectionWidget.tsx:41]: I think this is a standard pattern. See https://www.benmvp.com/blog/conditional-react-props-typescript/ and https://stackoverflow.com/questions/66917440/typescript-conditional-types-for-props.

Comment (pushyamig) [reply ccm_web/client/src/components/CreateSelectSectionWidget.tsx:41]: Thanks for additional info.

**Review by pushyamig — APPROVED** [review]
The change is functional and the create section widget is only given access to Admin and teacher users.
